fem/coefficient.cpp

```diff
@@ -170,6 +176,80 @@ void VectorGridFunctionCoefficient::Eval(
    GridFunc->GetVectorValues(T, ir, M);
 }
 
+GradientGridFunctionCoefficient::GradientGridFunctionCoefficient (
+   GridFunction *gf)
+   : VectorCoefficient((gf) ?
+                       gf -> FESpace() -> GetMesh() -> SpaceDimension() : 0)
+{
+   GridFunc = gf;
+}
+
+void GradientGridFunctionCoefficient::SetGridFunction(GridFunction *gf)
+{
+   GridFunc = gf; vdim = (gf) ?
+                         gf -> FESpace() -> GetMesh() -> SpaceDimension() : 0;
+}
+
+void GradientGridFunctionCoefficient::Eval(Vector &V, ElementTransformation &T,
+                                           const IntegrationPoint &ip)
+{
+   // Perhaps we should avoid altering T so cache its IntegrationPoint
+   const IntegrationPoint & ipT = T.GetIntPoint();
+   T.SetIntPoint(&ip);
+   GridFunc->GetGradient(T, V);
+   // Reset T's IntegrationPoint
+   T.SetIntPoint(&ipT);
+}
```

Comment:
The call to `GetGradient` will not alter `T`, so saving and restoring its integration point is redundant. Also, the call `T.SetIntPoint(&ip)` is unnecessary - this must be done by the caller already.

Reply:
I agree with you in part but I'm not comfortable making these assumptions.  We don't know how the calling routine might be written so we cannot be sure `T.SetIntPoint(&ip)` has been called.  However, I do know that `GridFunction::GetGradient()` will call `T.GetIntPoint();` so I feel obliged to make sure `T` contains a valid integration point.

I guess this is the connection between this PR and PR #593 that you had remembered and I had forgotten.  If PR #593 is merged then these issues disappear.

Reply:
It is a requirement for all `Eval` methods in the coefficient classes that take both `ElementTransformation &T` and `const IntegrationPoint &ip` as input, that `T.GetIntPoint()` is the same as `ip`. I believe that all places in MFEM where such calls happen, this behavior is ensured.

We should definitely explicitly spell out this requirement in the Doxygen documentation of these methods in the their base classes.

If you want to guard against calls that do not follow this requirement, it is better to add an `MFEM_ASSERT` here instead of setting and resetting the `IntegrationPoint` which breaks the caching for the Jacobian matrix and its inverse, determinant, etc.

Looking at the implementation of `GetGradient`, I realize that we can improve it - for example instead of calling `T.Jacobian()` and then calling `CalcInverse()` to invert it, we can simply call `T.InverseJacobian()` to take advantage of the caching.

Reply:
Right, placing an assertion of some sort was my first thought as well.  That was when I noticed that `T.GetIntPoint()` would lead to a segmentation fault if no point had been set.  Also, there was not reliable way to compare `IntegrationPoint` objects because they don't store their dimension and they don't always initialize their `y` or `z` values when used in lower dimensions.  The changes to make such checks possible are also in PR #593.

Reply:
I've removed the `GetIntPoint`/`SetIntPoint` calls in these three classes.